libs/datamodel/parser-database/src/attributes/default.rs

```diff
@@ -95,7 +85,84 @@ pub(super) fn visit_field_default<'ast>(
     }
 }
 
-fn validate_builtin_scalar_type_default(
+/// @default on composite type fields
+pub(super) fn visit_composite_field_default<'ast>(
+    args: &mut Arguments<'ast>,
+    field_data: &mut CompositeTypeField<'ast>,
+    ct_id: ast::CompositeTypeId,
+    field_id: ast::FieldId,
+    ctx: &mut Context<'ast>,
+) {
+    let value = match args.default_arg("value") {
+        Ok(value) => value,
+        Err(err) => return ctx.push_error(err),
+    };
+
+    let ast_model = &ctx.db.ast[ct_id];
+    let ast_field = &ast_model[field_id];
+
+    if ast_field.arity.is_list() {
+        return ctx.push_error(args.new_attribute_validation_error("Cannot set a default value on list field."));
+    }
+
+    if args.optional_arg("map").is_some() {
+        ctx.push_error(
+            args.new_attribute_validation_error("The `map` argument is not allowed on a composite type field."),
+        );
+    }
+
+    let default_attribute = args.attribute();
+
+    let mut accept = || {
+        let default_value = DefaultAttribute {
+            value: value.value,
+            mapped_name: None,
+            default_attribute,
+        };
+
+        field_data.default = Some(default_value);
+    };
+
+    // Resolve the default to a DefaultValue. We must loop in order to
+    // resolve type aliases.
+    let mut r#type = field_data.r#type;
+
+    loop {
+        match r#type {
+            ScalarFieldType::CompositeType(ctid) => {
+                validate_default_value_on_composite_type(ctid, ast_field, args, ctx);
+            }
+            ScalarFieldType::Enum(enum_id) => {
+                match value.value {
+                    ast::Expression::ConstantValue(enum_value, _) => {
+                        if ctx.db.ast[enum_id].values.iter().any(|v| v.name() == enum_value) {
+                            accept()
+                        } else {
+                            validate_invalid_default_enum_value(args, ctx);
+                        }
+                    }
+                    value => validate_invalid_default_enum_expr(value, args, ctx),
+                };
+            }
+            ScalarFieldType::BuiltInScalar(scalar_type) => {
+                validate_composite_builtin_scalar_type_default(scalar_type, value.value, accept, args, ctx)
+            }
+            ScalarFieldType::Alias(alias_id) => {
```

Comment:
We are removing the support for type aliases in the next major release. Maybe we don't need this anymore?

Reply:
Done. I added a validation that prevents type aliases from being used in composites and made that code path `unreachable!()`.